// react-css-collapse/lib/stories/App.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _addonActions = require('@storybook/addon-actions');

var _Collapse = require('../components/Collapse');

var _Collapse2 = _interopRequireDefault(_Collapse);

require('./style.css');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var App = function (_Component) {
  _inherits(App, _Component);

  function App(props) {
    _classCallCheck(this, App);

    var _this = _possibleConstructorReturn(this, (App.__proto__ || Object.getPrototypeOf(App)).call(this, props));

    _this.state = {
      index: 0
    };
    _this.setIndex = _this.setIndex.bind(_this);
    return _this;
  }

  _createClass(App, [{
    key: 'setIndex',
    value: function setIndex(index) {
      if (this.state.index !== index) {
        this.setState({ index: index });
      } else {
        this.setState({ index: null });
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      return _react2.default.createElement(
        'main',
        null,
        _react2.default.createElement(
          'section',
          null,
          this.props.elements.map(function (element, index) {
            return _react2.default.createElement(
              'div',
              { key: 'element-collapse-element-' + element.name },
              _react2.default.createElement(
                'button',
                {
                  onClick: function onClick() {
                    return _this2.setIndex(index);
                  },
                  style: { width: '100%' },
                  type: 'button'
                },
                element.name
              ),
              _react2.default.createElement(
                _Collapse2.default,
                _extends({
                  isOpen: _this2.state.index === index,
                  onRest: function onRest() {
                    return (0, _addonActions.action)('onRest');
                  }
                }, _this2.props.props),
                _react2.default.createElement(
                  'div',
                  { style: { background: 'lightpink', padding: '20px' } },
                  element.text
                )
              )
            );
          })
        ),
        _react2.default.createElement(
          'h3',
          null,
          'Other content...'
        )
      );
    }
  }]);

  return App;
}(_react.Component);

App.propTypes = {
  elements: _propTypes2.default.arrayOf(_propTypes2.default.shape({})).isRequired,
  props: _propTypes2.default.shape({
    className: _propTypes2.default.string,
    transition: _propTypes2.default.string
  }).isRequired
};

exports.default = App;
